### PR TITLE
18 role based access

### PR DESCRIPTION
**Pull request by woodwardmw**
This adds Supabase integration, along with a `/login` page, and "Log In" and "Log Out" buttons on the main UI page. Also javascript to show/hide these buttons, and to process user logins and logouts.

The `chatbot_auth_check_decorator` decorator function adds a `label` kwarg to `**kwargs`, which is a list of strings corresponding to the sources in the database that are listed for that user.

Everything seems to work, except the websocket disconnects, but I'm assuming that's something outside of this part of the code that will need to be fixed. Probably I'm not connecting to the database correctly.

### CODE REVIEW

**Comment by woodwardmw** [comment]
Note that it's now failing tests, because the chat won't function if the user isn't logged in.

File: README.md [hunk 1]
```diff
@@ -41,6 +41,7 @@ Environment variables and their default values,
 * `POSTGRES_DB_USER=admin`
 * `POSTGRES_DB_PASSWORD=secret`
 * `POSTGRES_DB_QUERY_LIMIT=10`
+* `DOMAIN=assistant.bible`
```
Comment (kavitharaju) [inline README.md:44]: Good to have the new ENV variables `SUPABASE_URL` and `SUPABASE_KEY`(not with value) listed here

File: app/core/auth/__init__.py [hunk 2]
```diff
@@ -10,40 +11,64 @@ def admin_auth_check_decorator(func):
...
+            'user_type', user_data.user.user_metadata.get('user_type')
+            ).limit(1).single().execute()
+        label = result.data.get('sources')
+
+        if not label:
+            raise PermissionException(
+            "Unauthorized access. User does not have access to any sources."
+            )
+
+        kwargs['label'] = label
```
Comment (kavitharaju) [inline app/core/auth/__init__.py:70]: As we are moving towards allowing multiple source types to be included in the same chat, in the auth also we need list of source-labels, not just one.

File: app/core/llm_framework/openai_langchain.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 '''Implemetations for lmm_framework interface using langchain'''
 import os
 from typing import List, Tuple
-from langchain.llms import OpenAI
+from langchain.chat_models import ChatOpenAI
```
Comment (kavitharaju) [inline app/core/llm_framework/openai_langchain.py:4]: I am curious here. Why the change? What are their differences?

File: app/routers.py [hunk 4]
```diff
@@ -388,3 +397,19 @@ async def get_source_tags(
...
+@router.get('/api/get-supabase-keys', response_model=schema.SupabaseKeys)
+def get_supabase_keys():
+    """Returns the supabase keys if set in environment variables"""
+    supabase_url = os.environ.get('SUPABASE_URL')
+    supabase_key = os.environ.get('SUPABASE_KEY')
+
+    if not supabase_url or not supabase_key:
+        return {"supabaseUrl": "", "supabaseKey": ""}
+
+    return {"supabaseUrl": supabase_url, "supabaseKey": supabase_key}
```
Comment (kavitharaju) [inline app/routers.py:410]: Is n't this API unsafe? Exposing the key to any user like this?

File: app/routers.py [hunk 5]
```diff
@@ -388,3 +397,19 @@ async def get_source_tags(
...
+
+@router.get("/modules/{module_name}")
+async def get_module(module_name: str):
+    """Returns the module file"""
```
Comment (kavitharaju) [inline app/routers.py:414]: Modules? Is this connected to supabase?

**Comment by kavitharaju** [comment]
I am curious what is "18 roles". Is there some seed data with predefined roles?

File: README.md (hunk 1 above)
Comment (woodwardmw) [reply README.md:44]: Done

File: app/core/llm_framework/openai_langchain.py (hunk 3 above)
Comment (woodwardmw) [reply app/core/llm_framework/openai_langchain.py:4]: I think I was getting an error running it, and it was telling me to make this change. But we could switch it back and see?

File: app/routers.py (hunk 5 above)
Comment (woodwardmw) [reply app/routers.py:412]: It was my javascript functions, to clean them up out of the html files. But not needed now, so deleted.

File: app/routers.py (hunk 4 above)
Comment (woodwardmw) [reply app/routers.py]: Now removed, so Supabase keys stay server-side.

**Comment by woodwardmw** [comment]
> I am curious what is "18 roles". Is there some seed data with predefined roles?

No - I think this was auto-generated from the branch name, which was 18-roles-... :smile:

File: app/core/auth/__init__.py [hunk 6]
```diff
@@ -60,14 +84,15 @@ async def wrapper(*args, **kwargs):
...
-        label = result.data.get('sources')
+        labels = result.data.get('sources')
 
-        if not label:
+        if not labels:
             raise PermissionException(
-            "Unauthorized access. User does not have access to any sources."
+            "Unauthorized access. User does not have access to any sources." 
+            # Or we could just give open access sources?
             )
```
Comment (kavitharaju) [inline app/core/auth/__init__.py]: Yes, if we plan to allow unregistered users and let them access the open-access sources, I guess we need to avoid raising errors here and in the other decorator where user token is checked. The code above may also have to be changed to expect None as user token.

File: app/routers.py [hunk 7]
```diff
@@ -398,18 +406,48 @@ async def get_source_tags(
...
+@router.post("/logout")
+async def logout(
+    ):
+    """Signs out a user"""
+    supabase_url: str = os.environ.get("SUPABASE_URL")
+    supabase_key: str = os.environ.get("SUPABASE_KEY")
+    supabase: Client = create_client(supabase_url, supabase_key)
+    supabase.auth.sign_out()
+
+    return {"next_url": f"http://{DOMAIN}/login"}
```
Comment (kavitharaju) [inline app/routers.py:436]: Having a {"message": "Logout successful"} also in the response will help UI apps to be sure about it.

File: app/routers.py [hunk 8]
```diff
@@ -398,18 +406,48 @@ async def get_source_tags(
...
+
+
+@router.post("/signup")
+async def signup(
+    email=Form(..., desc="Email of the user"),
+    password=Form(..., desc="Password of the user"),
+    ):
+    """Signs up a new user"""
```
Comment (kavitharaju) [inline app/routers.py:444]: Do we need more details about the user, like organization?
Where does the access allocation happen? Like which API lets us permit a user to access a label?

File: app/templates/chat-demo-postgres.html [hunk 9]
```diff
@@ -101,15 +101,28 @@ <h1 class="text-center">Chat Demo</h1>
...
+            console.log("Logging out");
+            event.preventDefault();
+            try {
+                const response = await fetch('/logout', {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json'
+                },
+                body: JSON.stringify({
+                })
```
Comment (kavitharaju) [inline app/templates/chat-demo-postgres.html:115]: The body params(email and password) are not seen here. Is this somehow connected to the form and will it be taken from there?

File: app/routers.py (hunk 7 above)
Comment (kavitharaju) [reply app/routers.py:436]: This could be done for other operations too. Like "Sign up succesfull", Signin successful" etc. Its not really required, but just a suggestion.

File: app/core/llm_framework/openai_langchain.py (hunk 3 above)
Comment (kavitharaju) [reply app/core/llm_framework/openai_langchain.py:4]: If tests are running locally then fine

File: app/templates/chat-demo-postgres.html (hunk 9 above)
Comment (woodwardmw) [reply app/templates/chat-demo-postgres.html:115]: Yes, I was surprised by that. But `supabase.auth.sign_out()` doesn't take any parameters. I think it must just identify the user from the locally stored accessToken, which we then delete.

File: app/routers.py (hunk 7 above)
Comment (woodwardmw) [reply app/routers.py]: Yep, I've just done that.

File: app/routers.py (hunk 8 above)
Comment (woodwardmw) [reply app/routers.py:450]: Yeah, good question. And the organization will need to verify it somehow, because we can't have a user just choosing their organization.
@joelthe1 What do you think?

File: app/core/auth/__init__.py (hunk 6 above)
Comment (woodwardmw) [reply app/core/auth/__init__.py]: OK yep, I've just done that now.

File: app/templates/chat-demo-postgres.html (hunk 9 above)
Comment (kavitharaju) [reply app/templates/chat-demo-postgres.html:115]: But how is that? If there are 10 user logged in at a time and we call supabase singout from server side, which user will be signed out.

Removing the token from localstorage may work, but the token also needs to be expired from supabase, right?

Comment (woodwardmw) [reply app/templates/chat-demo-postgres.html:115]: Hmmm... I've just googled this, and it seems that it isn't easy to log the user out on Supabase from the client side. 

But... does it matter if Supabase doesn't know you have logged out? A user is only logged in if they have a token. If you delete your client-side token, you're not logged in, and nor is anyone else in the world, since they don't have your token either. In order to be logged in, you, or anyone else, has to log in again and receive a new token.

Comment (kavitharaju) [reply app/templates/chat-demo-postgres.html:115]: I guess it may not matter much. 
But in the event the token is compromised, if the token doesn't get expired from the server, a hacker could access the systems using that token even after the user closes the app or clicks logout. Because from the supabase server side, the token would be still valid. But I dont know how critical an issue this is.

Comment (woodwardmw) [reply app/templates/chat-demo-postgres.html:115]: Yeah... We could maybe retrieve a new session, and not save the new token? https://supabase.com/docs/reference/python/auth-getsession

Comment (kavitharaju) [reply app/templates/chat-demo-postgres.html:115]: Looking through the docs makes me more confused. Will supabase allow multiple users to be logged in at the same time? Is it meant to be used from serverside or is it only client side?

File: app/core/auth/__init__.py [hunk 10]
```diff
@@ -76,21 +76,18 @@ async def wrapper(*args, **kwargs):
...
+            result = supabase.table('userTypes').select('''
+                    sources
+                    '''
+                ).eq(
+                'user_type', user_data.user.user_metadata.get('user_type')
+                ).limit(1).single().execute()
+            labels = result.data.get('sources')
 
         kwargs['label'] = labels
         # Proceed with the original function call and pass the sources to it
```
Comment (kavitharaju) [inline app/core/auth/__init__.py]: For this "public" label to work, all different kinds sources we expect to open up will need to have this one label "public".
How I had in my mind is some thing this,
You upload a bible, with label "ESV-Bible", a commentry with label, "Some-open-commentary" and such different types of resources with different labels.
Now in the access mapping table where we list all users and the labels they can access
1. either add these open labels to all users
2. or have a special row for "all_users" or "no-logged-in-user" etc and list these public labels there. 
I think such a approach will allow us to maintain the collections identity by unique labels and also move around things from public and propriteory if needed easily, just by changing the user-label table.

File: app/templates/chat-demo-postgres.html (hunk 9 above)
Comment (woodwardmw) [reply app/templates/chat-demo-postgres.html:115]: My impression is that it's designed to be used from client-side. Because I'm realizing, if we want to use other login methods (like Google Account logins, which should be easy), that will only work from client-side. Our current server-side implementation only works if someone is logging in using username and password.

File: app/routers.py [hunk 11]
```diff
@@ -386,3 +408,49 @@ async def get_source_tags(
...
+        "access_token": data.session.access_token
+        }
+
+
+@router.post("/logout")
+async def logout(
+    ):
+    """Signs out a user"""
+    supa.auth.sign_out()
+
```
[the remark this answers is not in the document]
Comment (woodwardmw) [reply app/routers.py:435]: @kavitharaju I've just tried this, and it seems to work fine.

File: app/core/auth/__init__.py [hunk 12]
```diff
@@ -2,48 +2,95 @@
...
+            raise ValueError("Access token is missing")
+        access_token_str = access_token.get_secret_value()
+        # Verify the access token using Supabase secret
+        try:
+            user_data = supa.auth.get_user(access_token_str)
+        except gotrue.errors.AuthApiError as e:
+            raise PermissionException("Unauthorized access. Invalid token.") from e
+
+        if user_data.user_metadata.get('user_type') != 'admin':
+            raise PermissionException("Unauthorized access. User is not admin.")
```
[the remark this answers is not in the document]
Comment (woodwardmw) [reply app/core/auth/__init__.py]: Yep, I've made that change now, so `user_data.user_metadata.user_types` (now plural) is an array. And `labels` is now the set of all labels that the user_types of the user can access.

**Review by kavitharaju — APPROVED** [review]
